File: safe_copy.go
package main

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"time"
)

func main() {
	// 1. Parse and validate arguments
	args := os.Args[1:]
	if len(args) < 2 || len(args) > 3 {
		fmt.Fprintf(os.Stderr, "Error: Invalid number of arguments.\n")
		fmt.Fprintf(os.Stderr, "Usage: %s <source_directory> <destination_directory> [overwrite]\n", os.Args[0])
		os.Exit(1)
	}

	srcDir := filepath.Clean(args[0])
	destDir := filepath.Clean(args[1])
	overwrite := false
	if len(args) == 3 && args[2] == "true" {
		overwrite = true
	}

	// 2. Setup log file
	logFileName := fmt.Sprintf("copy_log_%s.log", time.Now().Format("2006-01-02_15-04-05"))
	logFile, err := os.OpenFile(logFileName, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error creating log file: %v\n", err)
		os.Exit(1)
	}
	defer logFile.Close()

	logMessage := func(message string) {
		logLine := fmt.Sprintf("[%s] %s\n", time.Now().Format("2006-01-02 15:04:05"), message)
		fmt.Print(logLine)
		logFile.WriteString(logLine)
	}

	// 3. Pre-flight validations
	srcInfo, err := os.Stat(srcDir)
	if os.IsNotExist(err) || !srcInfo.IsDir() {
		fmt.Fprintf(os.Stderr, "Error: Source directory '%s' does not exist.\n", srcDir)
		os.Exit(1)
	}

	destInfo, err := os.Stat(destDir)
	if os.IsNotExist(err) || !destInfo.IsDir() {
		fmt.Fprintf(os.Stderr, "Error: Destination directory '%s' does not exist.\n", destDir)
		os.Exit(1)
	}

	// Read directory entries
	entries, err := os.ReadDir(srcDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading source directory: %v\n", err)
		os.Exit(1)
	}

	var files []os.DirEntry
	for _, entry := range entries {
		if !entry.IsDir() {
			files = append(files, entry)
		}
	}

	if len(files) == 0 {
		fmt.Fprintf(os.Stderr, "Error: Source directory '%s' is empty.\n", srcDir)
		os.Exit(1)
	}

	// 4. File Copy Logic
	logMessage(fmt.Sprintf("Starting file copy from '%s' to '%s'", srcDir, destDir))
	successCount, skipCount, failCount := 0, 0, 0

	for _, file := range files {
		filename := file.Name()
		srcFile := filepath.Join(srcDir, filename)
		destFile := filepath.Join(destDir, filename)

		if _, err := os.Stat(destFile); err == nil {
			if !overwrite {
				logMessage(fmt.Sprintf("Skipped: %s already exists in destination", filename))
				skipCount++
				continue
			}
		}

		// Perform safe low-level file write
		if err := copyFilePreservingAttributes(srcFile, destFile); err != nil {
			logMessage(fmt.Sprintf("Failed to copy: %s", filename))
			failCount++
		} else {
			logMessage(fmt.Sprintf("Copied: %s", filename))
			successCount++
		}
	}

	logMessage("File copy operation completed.")
	logMessage(fmt.Sprintf("Summary -> Successful: %d | Skipped: %d | Failed: %d", successCount, skipCount, failCount))

	if failCount > 0 {
		os.Exit(1)
	}
}

func copyFilePreservingAttributes(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()

	if _, err = io.Copy(out, in); err != nil {
		return err
	}

	srcInfo, err := os.Stat(src)
	if err != nil {
		return err
	}

	// Preserve permissions
	if err := os.Chmod(dst, srcInfo.Mode()); err != nil {
		return err
	}

	// Preserve timestamps
	return os.Chtimes(dst, time.Now(), srcInfo.ModTime())
}

File: safe_copy.php
#!/usr/bin/env php
<?php

// 1. Parse and validate arguments
if ($argc < 3 || $argc > 4) {
    fwrite(STDERR, "Error: Invalid number of arguments.\n");
    fwrite(STDERR, "Usage: php {$argv[0]} <source_directory> <destination_directory> [overwrite]\n");
    exit(1);
}

$srcDir = rtrim($argv[1], '/');
$destDir = rtrim($argv[2], '/');
$overwrite = isset($argv[3]) && strtolower($argv[3]) === 'true';

// 2. Setup timestamped log file
$logFile = "copy_log_" . date('Y-m-%d_H:i:s') . ".log";

function logMessage($message, $logFile) {
    $timestamp = date('Y-m-%d H:i:s');
    $logLine = "[$timestamp] $message\n";
    echo $logLine;
    file_put_contents($logFile, $logLine, FILE_APPEND);
}

// 3. Pre-flight validations
if (!is_dir($srcDir)) {
    fwrite(STDERR, "Error: Source directory '{$srcDir}' does not exist.\n");
    exit(1);
}

if (!is_readable($srcDir)) {
    fwrite(STDERR, "Error: Missing read permissions on source directory '{$srcDir}'.\n");
    exit(1);
}

if (!is_dir($destDir)) {
    fwrite(STDERR, "Error: Destination directory '{$destDir}' does not exist.\n");
    exit(1);
}

if (!is_writable($destDir)) {
    fwrite(STDERR, "Error: Missing write permissions on destination directory '{$destDir}'.\n");
    exit(1);
}

// Read and isolate files only
$allItems = scandir($srcDir);
$files = [];
foreach ($allItems as $item) {
    if ($item === '.' || $item === '..') continue;
    if (is_file($srcDir . '/' . $item)) {
        $files[] = $item;
    }
}

if (count($files) === 0) {
    fwrite(STDERR, "Error: Source directory '{$srcDir}' is empty.\n");
    exit(1);
}

// 4. File Copy Logic
logMessage("Starting file copy from '{$srcDir}' to '{$destDir}'", $logFile);
$successCount = 0;
$skipCount = 0;
$failCount = 0;

foreach ($files as $file) {
    $srcFile = $srcDir . '/' . $file;
    $destFile = $destDir . '/' . $file;

    if (file_exists($destFile)) {
        if (!$overwrite) {
            logMessage("Skipped: $file already exists in destination", $logFile);
            $skipCount++;
            continue;
        }
    }

    if (copy($srcFile, $destFile)) {
        // Touch to preserve modification time mapping
        touch($destFile, filemtime($srcFile));
        logMessage("Copied: $file", $logFile);
        $successCount++;
    } else {
        logMessage("Failed to copy: $file", $logFile);
        $failCount++;
    }
}

logMessage("File copy operation completed.", $logFile);
logMessage("Summary -> Successful: $successCount | Skipped: $skipCount | Failed: $failCount", $logFile);

exit($failCount > 0 ? 1 : 0);
